[Andre Wong Zhi Hua] ip

AndreWongZH opened the pull request:
I have rebased my earlier commits from level-0 to level-3 so the time of commit is inaccurate.

AndreWongZH committed:
Add skeletal version of Duke

Lets add in code to make the program greet the user
and then exits.

AndreWongZH committed:
Add user interactions to program

Duke class: greet and say bye to user

Duke class needs more functionality to interact with users. It does
not allow users to enter any inputs.

Lets add in the functionality of echoing commands that
users has entered.
The program will exit when user enters "bye" to prevent it from
running indefinitely.

AndreWongZH committed:
Add the ability to store text by user

Duke class: echo messages entered by user

We need to have the ability to store texts entered by user.

Lets create a new class called Items to keep a record of texts
entered by user up to a maximum of 100 texts.

It has a public method addToList to add a text entered by user.
It has a public method printList to display all the text entered.

AndreWongZH committed:
Add the ability to mark tasks as done

Items class: keep track of the list of user text inputs

We need a way for the Item class to keep track of the status of the
task.

Lets rename the Items class to be called TaskManager and it's
methods to better reflect the purpose of the class.
Added setTaskDone method to TaskManager class.

I also added a new class called Task to keep track of the task
description and whether if it is done or not.
The Task class has a public method getTaskStatus to return the done
status and description of the task as a string.

AndreWongZH committed:
Refactor code to comply with coding standards

Classes Duke, Task and TaskManager currently comply with the coding
standards.

This can affect the readability of the code.

Lets update all the classes and public methods to include a javadoc
and ensure that scope of the variables are defined at the smallest
scope possible.
I also ensured that whitespaces are used in statements to improve
readability.

AndreWongZH committed:
Add support for tracking todos, deadlines and events

Task class: records the description and isDone boolean
TaskManager class: keep tracks of the list of task class

The task class needs to be split into todos, deadlines, and events.
Each class will have additional attributes: deadlines will have a
by attribute and events will have an at attribute.

Lets have our Task class be the parent class and todo, deadline and
event class will inherit from it.
We will also have a new method splitItem for the TaskManager class
to handle the separation of user input into task and date/time.
We will also have a TaskType enumeration as currently there is only
todo, deadline and event type of tasks, more can be added in the
future.

Using inheritance here is preferred because all three class types
have description and isDone attributes and this reduces repetition.

AndreWongZH committed:
Add new text automation testing

Lets create tests for our program using I/O redirection technique.

This is because in the future we can be notified if any of our test
failed due to new changes being made.

AndreWongZH committed:
Refactor Duke class for readability

Duke class: extract methods from Duke main class

Duke class now have varying levels of abstraction within the main
method.

Lets make the main method in Duke class to have the same level of
abstraction by extracting out lines of code into a method.
I also added static string constants for greetings and named
constants to avoid magic numbers.

Have the same level of abstraction allows us to adhere to SLAP
principle.
Not having magic numbers increases readability

AndreWongZH committed:
Refactor Task, Todo, Deadline and Event class for readability

There are many single strings that are scattered about in the print
statements, causing code to be hard to read

Lets have named constants for opening and closing of strings to be
printed out. We also have constants for the tick and cross icons.
We also have constants for task tags.

AndreWongZH committed:
Refactor TaskManager class

Extracted methods from other methods to ensure SLAP principle is
maintained.
Magic literals are extracted out as named constants.
Overload deadline and event class's constructors to make it possible
to pass in one argument instead, improves readability.
Added comments to private methods.

felixhalim commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,119 @@
+            task = new Event(descAndTime);
+            break;
+        default:
+            throw new IllegalStateException("Unexpected value: " + taskType);
```
I like the way you implement Java error checking by throw exception.

AndreWongZH replied:
Thanks!

felixhalim commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,119 @@
+        case EVENT:
+            index = inputText.indexOf(AT_LITERAL);
+            break;
+        }
```
Perhaps you can add "default" case to adhere with the Code Standard

AndreWongZH replied:
Thanks for pointing out this violation, I might have missed it.

felixhalim reviewed (commented):
Please review my comments, overall, it's good code quality

QX-CHEN commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,75 @@
+        return in.nextLine();
+    }
+
+    private static boolean checkIfUserQuits(String userInput) {
```
Perhaps the function name could change to shouldQuit or isDone? When the input is "bye", the program should quit.

QX-CHEN commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,119 @@
+    /**  splits user input into descriptions and date/time */
+    private String[] splitItem(TaskType taskType, String inputText) {
+        int index = 0;
+        String[] descAndTime = new String[MAX_INPUT_PARAMS];
```
Perhaps can replace desc with description? Took me a second to realise it's description.

AndreWongZH replied:
Alright, thanks for pointing it out to me!

QX-CHEN reviewed (commented):
Overall, great work! The code is clear and easy to read.

AndreWongZH committed:
Fix TaskManager class to follow code standards

Lets add a default case to adhere with the code standards for the
splitItem method.
I renamed the descAndTime variable to taskParameters for better
readability.

AndreWongZH committed:
Add handling of exceptions for Duke class

Duke class does not handle invalid commands and empty commands which
requires parameters.

Lets create a new method under the Duke class to extract out the
command. If no command is extracted, method will throw an
IllegalCommandException.

Lets also have a new method to extract out the parameters after the
commands. If no parameters are entered, the method will throw an
IndexOutOfBoundsException.

AndreWongZH committed:
Add handling of exceptions for TaskManager class

TaskManager does not handle invalid or missing user parameters.

Lets throw a MissingTaskLiteralException if the program does not
detect a /at or /by and prompts the user to re enter a new command.

For setting task to be done, we also handled cases where parameter
passed in is not a integer or the integer is out of range of the
available list.

AndreWongZH committed:
Fix exceptions in Duke and TaskManager class

Duke class: Bye command does not exit the program.
TaskManager class: Does not handle situation where list if full.

Lets add in a new BYE enum for extractCommand method to check for
bye command. I also extracted out the error messages as static
string constants. I also changed checking of user input to use
.startsWith() rather than .contains() as it prevents inputs like
"event done" to give a wrong error.

Lets also handle IndexOutOfBoundsException when the program tries to
add a task to a full list.

AndreWongZH committed:
Fix and add more test cases

Updated runtext.bat file to complile all my java files.
Added in more test cases where error in user input occur and
handled.

AndreWongZH committed:
Merge branch 'branch-Level-5' into master

AndreWongZH committed:
Divide classes into packages

Currently the classes do not belong to any packages.

Lets group the classes into 2 different packages under the duke
package, the command and task package.

I also extracted out methods in Duke class that pertains to commands
and made a new class called CommandManager.

I also updated the runtest.bat file to ensure that compile path is
correct.

AndreWongZH committed:
Merge branch 'branch-A-Packages'

AndreWongZH committed:
Add command delete to command class

The command class does not support deleting tasks.

Lets add a new command enum for delete and support detecting the
delete command. Also moved variables taskManager, parameters and
command from main method to class variable as these variables
are scoped under the entire class.

AndreWongZH committed:
Add delete function to TaskManager

The TaskManager class is unable to delete tasks from the task list.

Lets first convert the array we have into an ArrayList and replace
all the corresponding array methods with ArrayList methods (such as
.get() and .add()).
Next I added a new method, deleteTask(). I also removed exception
when the ArrayList is full.
Lastly, I added an interface to the taskManager so that it knows to
override the public methods for better abstraction.

AndreWongZH committed:
Add test for delete command

Added test for deleting command and added interface to list of
classes to compile.
Updated test cases.

AndreWongZH committed:
Add ability to load data from disk

Data is current not saved and information will be gone once the
programm exits.

Lets create a FileManager class that can load data from
./data/duke.txt. The Path type is used to pass to the scanner to
ensure cross platform file paths are compatible.

Once data is retrived, we then pass it to the DataParser class to
convert the String text to Task Objects to be added to the
taskManager.

I have also overloaded the Task constructor so that it can accept
isDone boolean.

AndreWongZH committed:
Add ability to save task to hard disk

Currently we can only load data from our drive but unable to write
to duke.txt.

Lets create a method under FileManager class to write data to
duke.txt. We also created a FileWritable interface for the task
class for it to convert the object to a string to be saved.

AndreWongZH committed:
Merge branch 'branch-Level-6'

AndreWongZH committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/duke/task/TaskManager.java

Changed from array to ArrayList, removed taskCount integer.

Overloaded the constructor for Taskmanager so that it can feed in
previous data from disk.

Let FileManager class handle all the exceptions for file errors.

Program will auto create data file if ./data/duke.txt is not found.

Updated covertToData method of the Task class for better readbility.

AndreWongZH committed:
Update test for loading data

text-ui-test was not working properly due to the duke.txt file not
being deleted after every file compare.

Lets remove the duke.txt file so that every new test will have an
empty duke.txt file.
Updated compile files for test.

AndreWongZH committed:
Handle file read format error

The DataParser class might give errors when the format of duke.txt
is not correct.

Lets handle the exception by creating FileCorruptedException to
throw and this is handled by the FileManager. The FileManager will
then delete the corrupted file and then recreate a new duke.txt.

Added a new test case to test when duke.txt is corrupted

AndreWongZH committed:
Add ability to create data directory

The FileManager does not check if data directory exist and that is
giving the program errors.

Lets get the FileManager to check if data directory exist and then
create the directory is it does not exist.

Fixed line spacing for error messages and updated the test expected
results.

joelczk commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,5 @@
+package duke.command;
+
+public enum Command {
+    LIST, DONE, TODO, DEADLINE, EVENT, BYE, DELETE
+}
```
Gd attempt at using enums!

joelczk commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @param done boolean status of the task.
+     */
+    public void setDone(boolean done) {
```
`Boolean` variables should be named to sound like booleans(i.e. isDone, isSet etc.) Please consider changing the name of your boolean variable.

AndreWongZH replied:
Thanks for pointing it out to me!

joelczk commented on src/main/java/duke/task/TaskAction.java:
```diff
@@ -0,0 +1,8 @@
+package duke.task;
+
+public interface TaskAction {
+    void addTask(TaskType taskType, String inputText);
+    void printAllTasks();
+    void setTaskDone(String inputText);
+    void deleteTask(String inputText);
+}
```
Gd attempt at using interface!

joelczk commented on src/main/java/duke/task/TaskManager.java:
```diff
@@ -0,0 +1,211 @@
+     */
+    @Override
+    public void printAllTasks() {
+        if (tasks.size() == 0) {
```
Try to reduce the use of magic numbers. Maybe try to declare it in a constant variable?

AndreWongZH replied:
Yep will do!

joelczk commented on src/main/java/duke/task/TaskType.java:
```diff
@@ -0,0 +1,5 @@
+package duke.task;
+
+public enum TaskType {
+    TODO, DEADLINE, EVENT
+}
```
Gd use of enums!

joelczk reviewed (commented):
Generally quite a clean piece of work. Take note of the comments and make the relevant changes.

AndreWongZH committed:
Fix coding standards violations

Fixed some parts of the code that have magic numbers and
boolean parameters are not named correctly.

AndreWongZH committed:
Update .gitignore file

Ignore the manifest file created by build artifacts and as well as
ignoring Actual outputs created by the runtest.bat. Ignored the data
directory created during test as well.

AndreWongZH committed:
Refactor the CommandManager class

CommandManager class now handles add, delete, done, list, and bye
commands as 1 class.

Lets extract out the individual commands into its indiviual classes
for better readbility and more OOP. As such, we have Add, Bye, Done,
List and Delete class that will inherit from the Command class. Each
class will override the Command class's execute command.

AndreWongZH committed:
Refactor CommandManager into CommandParser

Moved extract command and extract parameters methods into the new
CommandParser class and this is created in the Duke class. The
parseCommand will handle the extraction of command and parameters,
before executing the command.

Also created a new GreetCommand class to call the greeting.

AndreWongZH committed:
Refactor TaskManager class

Previously, the TaskManager class handles converting the parameters
into taskNumber, description and dateTime. This makes the class do
too much.

Lets move the parsing of parameters to ParameterParser class, which
its purpose is to parse the userinput into integers for delete and
done command, and description and datetime for add command.

The ParameterData class stores the data parsed by the
ParameterParser for better readability.

The TaskAction interface and Command classes are updated to follow
this new implementation.

AndreWongZH committed:
Refactor FileManager class

previously Filemanager class methods are all static.

Lets create a Filemanager class that handles all the loading and
writing of data through an instance of the class.

The code inside the FileManager class are also extracted out to
ensure SLAP principle is followed. This give more readability.

Fix bug due to missing break in switch statement.

AndreWongZH committed:
Add FileAction interface

Added FileAction interface for FileManager class to show which
methods are open to public.

AndreWongZH committed:
Refactor output code to Ui classes

Right now all the printing of user output is everywhere in my code.

Lets create a ui package to handle all the printing of output.
This is splited up into CommandUi, FileUi, InputUi, and TaskUi,
which all inherits the Ui abstract class.

Each class will handle the printing of error and success cases and
printing information.

The InputUi class will also be incharge of reading in the user input
and trimming it.

AndreWongZH committed:
Tune exception handling

Some of the failure cases are printing too much info. Updated
them to only catch at different areas of the code.

Updated .gitignore file to prevent pushing ACUTAL.TXT.

AndreWongZH committed:
Fix runtest.bat to be runnable

runtest.bat was failing due to the refactoring changes.
Updated the runtest.bat file to compile all the new java classes and
update the expected outputs.

AndreWongZH committed:
Add datetime for Duke

Right now we are only passing the datetime parameter of deadline and
event as a string.

Lets teach duke to understand more on date and time and we first
convert all the string parameters of datetime into a LocalDateTime
object along with the other classes that uses the datetime
parameter.

AndreWongZH committed:
Add a datetime parser class

Now we can store our datetime as LocalDateTime objects but are
no able to parse from our user input yet.

Lets create a DateTimeParser class to handle the user input.
We also have created an enum of all valid datetime format that my
program accepts.
The DateTimeParser will try every single format to see if the user
input fits any of the valid format. If there are no valid format,
it will throw a DateTimeFormatException and this will be handled
by the CommandUi class, informing user that datetime format is
wrong.

AndreWongZH committed:
Add ability to print datetime

Now we can convert user inputs into datetime objects, we need a way
to display the time back to them.

Lets have a DateTimePrintable interface that the Event and Deadline
class will implement so that they can print out the datetime
pattern based on a formatter.
I have also updated the DateTimeFormat Enum to include more formats
the user inputs we can accept.
When writing the task to file, the time format used is also clean
and easy for the user to change manually.

AndreWongZH committed:
Add ability to sanitize datetime inputs

Right now user need to type Jan with front capitalized and AM/PM in
all caps in order for the program to register the correct datetime.

Lets sanitize the user inputs such that it will auto lowercase the
input and then proceed to replace it with the correct format.

E.g JAN -> jan -> Jan and Am -> am -> AM

AndreWongZH committed:
Add ability for list command to filter by date

Lets allow user to be able to filter their todo list by date.
I have added a /on literal to the list command and followed by the
date to query for. This will return a list of all tasks on that
date. If no result is found, an error will be returned to the user.

To simplfy the command further, user can enter list /on today to
get the list of all the tasks on today's date.

All todos will have a date of today.

AndreWongZH committed:
Add ability to find task by keyword

Lets create a new command, ListCommand that takes in a filter string
and then prints out all the related task.

AndreWongZH committed:
Merge pull request #1 from AndreWongZH/branch-Level-8

Merge branch level 8 to master branch

AndreWongZH committed:
Add JavaDoc comments for Ui classes

No comments added if the methods are pretty obvious at what they
are doing.

AndreWongZH committed:
Add JavaDoc comments for task classes

Added JavaDoc to all classes and methods under task packages.

AndreWongZH committed:
Add JavaDoc comments for command classes

Added JavaDoc to all classes and methods under command package.

AndreWongZH committed:
Add JavaDoc comments for storage classes

Added JavaDoc to all classes and methods under storage package.

AndreWongZH committed:
Add JavaDoc comments for parser classes

Added JavaDoc to all classes and methods under parser package.

AndreWongZH committed:
Add JavaDoc comments for Duke class.

AndreWongZH committed:
Merge branch 'master' into branch-Level-9

# Conflicts:
#	src/main/java/duke/parser/CommandParser.java
#	src/main/java/duke/parser/ParameterParser.java
#	src/main/java/duke/task/Task.java

Refactored the code in the TaskManager class.

AndreWongZH committed:
Merge pull request #2 from AndreWongZH/branch-Level-9

Merge branch level 9 into master branch

AndreWongZH committed:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/parser/DataParser.java
#	src/main/java/duke/parser/ParameterData.java
#	src/main/java/duke/parser/ParameterParser.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskAction.java
#	src/main/java/duke/task/TaskManager.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/ui/TaskUi.java

Updated all the new comments on findCommand, dateTimeParser class.

AndreWongZH committed:
Merge pull request #3 from AndreWongZH/branch-A-JavaDoc

Merge Branch A JavaDoc into master branch

AndreWongZH committed:
Update find command

Right now my list command takes in a /on parameter to list by date.
While my find command takes in a user string to search description
by string.

Lets move the /on parameter from the list command to the find
command. So now the find command can either search by keyword or by
date or both.

Lets also add in the ability to search before or after by using the
/bf and /af literal instead of the /on literal. This way they can
search for any tasks before or after the given date.

AndreWongZH committed:
Update done command

The done command only lets you to set the task to done. It does not
allow user to undone any task.

Lets improve on the done command by setting the done command to
toggle the task's done attribute.
So done 1 -> set task to done -> done 1 -> set task to undone.

AndreWongZH committed:
Update user input interface

The input for user commands ui seems boring.

Lets enhance by adding a header to the output so that it shows
the current number of todos, deadlines and events the user
have in their tasks.

Lets inform users the total number of tasks saved when the user
wants to exit the program.

AndreWongZH committed:
Replace tick and cross for done.

Since the tick and cross has problems showing up in other OS, I
have changed it to using Y and X instead.

AndreWongZH committed:
Update text-ui-test

Update all the compiled class file paths under runtest.bat.
Update all the test cases for new commands and new features.
Fix parsing datetime problem by trimming user input.

AndreWongZH committed:
Update user guide

Update the user guide with all the commands, with the format and
examples. Currently leaving the sample output empty. Created a
command summary table at the end of the user guide.

AndreWongZH committed:
Set theme jekyll-theme-cayman

AndreWongZH committed:
Update user guide

Updated user guide to fix table and cayman config files.

AndreWongZH committed:
Update readme file for website

Add in a content page, images and more sections to the readme file.

AndreWongZH committed:
Update readme document

Add in pictures for all the expected outputs and also quick start.

AndreWongZH committed:
Make find command case insensative

Add in additional functionaility for find to search keywords to be
case insensative.

AndreWongZH committed:
Update ui for find command

The heading for find command seems too similar to list command.

Lets have the find command have its own unique heading.

AndreWongZH committed:
Updated readme document

Add in the date formatting requirements inside the user guide.

AndreWongZH committed:
Fix corrupted file problem

Program will throw an error if the boolean parsed is not 0 or 1
during the reading of duke.txt.

Updated readme document to include FAQ.

AndreWongZH committed:
Refactor code to adhere with coding standards

AndreWongZH committed:
Update for deadline and event to get date

Add the function for deadline and event to read in today after
\by or \on.

Calling today for deadline and event will set date to current date
and at 2359 time.

AndreWongZH committed:
Add ability to reject add tasks

Lets make the program check with the tasks list to see if there are
any duplicates of the same task before adding a new task in.

It will check the description for todos, and description and
datetime for event and deadline.

Updated testing outcomes and userguide as well.

AndreWongZH committed:
Add ability for list to sort

Lets add additional parameter, asc and desc to the list command so
that user can sort by ascending or descending order.

Update user doc to relect this change as well as text-ui-test.

AndreWongZH committed:
Update jar version to 2.0

AndreWongZH committed:
Update list by order output

Change the list output via asc and desc to be an arrow than a
number.

AndreWongZH committed:
Add help command

Lets give a command to help users easilly refer to all the available
command formats incase they forget.

AndreWongZH committed:
Update text-ui-test
